src/examples/4626-router/Generic4626Router.sol

```diff
@@ -0,0 +1,229 @@
+    }
+
+    /// @inheritdoc BaseHook
+    function getHookPermissions() public pure override returns (Hooks.Permissions memory) {
+        return Hooks.Permissions({
+            beforeInitialize: true, // Validate settings
+            beforeAddLiquidity: true, // Disallow adding liquidity
+            beforeSwap: true, // Handle wrapping/unwrapping
+            beforeSwapReturnDelta: true, // Async Swap via the vault
+            afterSwap: false,
+            afterInitialize: false,
+            beforeRemoveLiquidity: false,
+            afterAddLiquidity: false,
+            afterRemoveLiquidity: false,
+            beforeDonate: false,
+            afterDonate: false,
+            afterSwapReturnDelta: false,
+            afterAddLiquidityReturnDelta: false,
+            afterRemoveLiquidityReturnDelta: false
+        });
+    }
+
+    function _beforeInitialize(
+        address,
+        PoolKey calldata poolKey,
+        uint160
+    ) internal view override returns (bytes4) {
+        if (poolKey.fee != 0) {
+            revert Generic4626Router__InvalidPoolFee();
+        }
+
+        PoolId poolId = poolKey.toId();
+        PoolDetails memory details = poolDetails[poolId];
+
+        if (!details.isInitialized) {
+            // We enforce pool initialization via the hook, this way we can
+            // ensure that the pool is initialized with the correct parameters
+            revert Generic4626Router__NotAllowed();
+        }
+
+        return IHooks.beforeInitialize.selector;
+    }
+
+    function _beforeAddLiquidity(
+        address,
+        PoolKey calldata,
+        IPoolManager.ModifyLiquidityParams calldata,
+        bytes calldata
+    ) internal pure override returns (bytes4) {
+        revert Generic4626Router__NotAllowed();
+    }
+
+    function _beforeSwap(
+        address,
+        PoolKey calldata poolKey,
+        IPoolManager.SwapParams calldata params,
+        bytes calldata
+    )
+        internal
+        override
+        returns (bytes4 selector, BeforeSwapDelta swapDelta, uint24 lpFeeOverride)
+    {
+        PoolId poolId = poolKey.toId();
+        PoolDetails memory details = poolDetails[poolId];
+
+        bool isExactInput = params.amountSpecified < 0;
+
+        // This following section is hard to follow, I initially wrote this with currency0 and currency1
+        // but it was truly impossible to follow. Ended up storing the vault and underlying.
+        // Although, I think I got a hang of it so might go back
+        if (params.zeroForOne == details.wrapDirection) {
+            uint256 inputAmount = isExactInput
+                ? uint256(-params.amountSpecified)
+                : _getUnderlyingForShares(details.vault, uint256(params.amountSpecified));
+
+            CurrencySettler.take(
+                Currency.wrap(address(details.underlying)),
+                poolManager,
+                address(this),
+                inputAmount,
+                false
+            );
```

Comment:
if we set the directive `using CurrencySettler for Currency` we get really nice syntactic sugar:

```solidity
details.underlying.take(poolManager, address(this), underlyingAmount, false);
```

Reply:
I personally do not like using those, largely because they make debugging harder. That said, I think this makes sense purely because it's a syntax that v4 uses everywhere so might as well.